src/dotty/tools/dotc/transform/DropEmptyCompanions.scala

```diff
@@ -0,0 +1,98 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import DenotTransformers.SymTransformer
+import Phases.Phase
+import Contexts.Context
+import Flags._
+import Symbols._
+import SymDenotations.SymDenotation
+import ast.Trees._
+import collection.mutable
+import Decorators._
+import NameOps._
+import TreeTransforms.MiniPhaseTransform
+import dotty.tools.dotc.transform.TreeTransforms.TransformerInfo
+
+/** Remove companion objects that are empty
+ *  Lots of constraints here:
+ *  1. It's impractical to place DropEmptyCompanions before lambda lift because dropped
+ *     modules can be anywhere and have hard to trace references.
+ *  2. DropEmptyCompanions cannot be interleaved with LambdaLift or Flatten because
+ *     they put things in liftedDefs sets which cause them to surface later. So
+ *     removed modules resurface.
+ *  3. DropEmptyCompanions has to be before RestoreScopes.
+ *  The solution to the constraints is to put DropEmptyCompanions between Flatten
+ *  and RestoreScopes and to only start working once we are back on PackageDef
+ *  level, so we know that all objects moved by LambdaLift and Flatten have arrived
+ *  at their destination.
+ */
+class DropEmptyCompanions extends MiniPhaseTransform { thisTransform =>
+  import ast.tpd._
+  override def phaseName = "dropEmpty"
+  override def runsAfter: Set[Class[_ <: Phase]] = Set(classOf[Flatten])
+
+  override def transformPackageDef(pdef: PackageDef)(implicit ctx: Context, info: TransformerInfo) = {
+
+    /** Is `tree` an empty companion object? */
+    def isEmptyCompanion(tree: Tree) = tree match {
+      case TypeDef(_, impl: Template) if tree.symbol.is(SyntheticModule) &&
+        tree.symbol.companionClass.exists &&
+        impl.body.forall(_.symbol.isPrimaryConstructor) =>
+        println(i"removing ${tree.symbol}")
```

Comment:
It's already removed in another PR of mine. So it should go away by itself
soon.

On Thu, Feb 18, 2016 at 11:23 AM, Dmitry Petrashko <notifications@github.com
> wrote:

> In src/dotty/tools/dotc/transform/DropEmptyCompanions.scala
> <https://github.com/lampepfl/dotty/pull/1075#discussion_r53295488>:
>
> > + *  level, so we know that all objects moved by LambdaLift and Flatten have arrived
> > + *  at their destination.
> > + */
> > +class DropEmptyCompanions extends MiniPhaseTransform { thisTransform =>
> > +  import ast.tpd._
> > +  override def phaseName = "dropEmpty"
> > +  override def runsAfter: Set[Class[_ <: Phase]] = Set(classOf[Flatten])
> > +
> > +  override def transformPackageDef(pdef: PackageDef)(implicit ctx: Context, info: TransformerInfo) = {
> > +
> > +    /** Is `tree` an empty companion object? */
> > +    def isEmptyCompanion(tree: Tree) = tree match {
> > +      case TypeDef(_, impl: Template) if tree.symbol.is(SyntheticModule) &&
> > +        tree.symbol.companionClass.exists &&
> > +        impl.body.forall(_.symbol.isPrimaryConstructor) =>
> > +        println(i"removing ${tree.symbol}")
>
> println left in code.
> I'll make a pr to remove it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/pull/1075/files#r53295488>.
>



-- 
Martin Odersky
EPFL